src/main/kotlin/it/pagopa/wallet/services/WalletService.kt

```diff
@@ -529,36 +518,71 @@ class WalletService(
                 }
             )
 
-    private fun updateServiceList(
-        wallet: Wallet,
-        dataService: Pair<ServiceName, ServiceStatus>
-    ): Wallet {
-        val updatedServiceList = wallet.applications.toMutableList()
-        when (
-            val index =
-                wallet.applications.indexOfFirst { s -> s.name.name == dataService.first.name }
-        ) {
-            -1 ->
-                updatedServiceList.add(
-                    Application(
-                        ServiceId(UUID.randomUUID()),
-                        ServiceName(dataService.first.name),
-                        ServiceStatus.valueOf(dataService.second.name),
-                        Instant.now()
-                    )
-                )
-            else -> {
-                val oldWalletService = updatedServiceList[index]
-                updatedServiceList[index] =
-                    Application(
-                        oldWalletService.id,
-                        oldWalletService.name,
-                        ServiceStatus.valueOf(dataService.second.name),
-                        Instant.now()
-                    )
+    fun updateWalletServices(
+        walletId: UUID,
+        servicesToUpdate: List<Pair<ServiceName, ServiceStatus>>
+    ): Mono<LoggedAction<WalletServiceUpdateData>> {
+        return walletRepository
+            .findById(walletId.toString())
+            .switchIfEmpty { Mono.error(WalletNotFoundException(WalletId(walletId))) }
+            .flatMap { wallet ->
+                val walletApplications = wallet.applications.associateBy { it.name }.toMutableMap()
+
+                servicesToUpdate
+                    .toFlux()
+                    .flatMap { (serviceName, requestedStatus) ->
+                        serviceRepository.findById(serviceName.name).map {
+                            Triple(it, serviceName, requestedStatus)
+                        }
+                    }
+                    .reduce(
+                        Pair(
+                            mutableMapOf<ServiceName, ServiceStatus>(),
+                            mutableMapOf<ServiceName, ServiceStatus>()
+                        )
+                    ) {
+                        (servicesUpdatedSuccessfully, servicesWithUpdateFailed),
+                        (service, serviceName, requestedStatus) ->
+                        val serviceGlobalStatus = ServiceStatus.valueOf(service.status)
+                        val walletApplication = walletApplications[serviceName.name]
+
+                        if (
+                            ServiceStatus.canChangeToStatus(
```

Comment:
Coverage for all pr code is really good. But You could improve adding specific test for this method.

Reply:
Fixed with d94e391